core/chain_processor.go

```diff
@@ -0,0 +1,176 @@
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package core implements the Ethereum consensus protocol.
+package core
+
+import (
+	"sync"
+	"time"
+)
+
+// ChainProcessor is an external process that creates auxiliary data structures
+// using the canonical chain as an input. Its callback function NewHead is called every
+// time a new head is added to the chain or it gets rolled back. The callback function
+// should never block because it is called under the chain's mutex lock so that the
+// structures can stay consistent with the canonical chain.
+// The same interface can be used for chaining processors (one using the output of the
+// other).
+type ChainProcessor interface {
+	NewHead(headNum uint64, rollBack bool)
+}
+
+// ChainSectionProcessor is a ChainProcessor that does a post-processing job for
+// equally sized sections of the canonical chain (like BlooomBits and CHT structures).
+// Further child ChainProcessors can be added which use the output of this section
+// processor. These child processors receive NewHead calls only after an entire section
+// has been finished or in case of rollbacks that might affect already finished sections.
+type ChainSectionProcessor struct {
+	backend                        ChainSectionProcessorBackend
+	sectionSize, confirmReq        uint64
+	stop                           chan struct{}
+	updateChn                      chan uint64
+	lock                           sync.Mutex
+	calcValid                      bool
+	procWait                       time.Duration
+	stored, calcIdx, lastForwarded uint64
+	childProcessors                []ChainProcessor
+}
+
+// ChainSectionProcessorBackend does the actual post-processing job.
+// GetStored and SetStored are called under the canonical chain lock and should
+// not block. Process is called without a chain lock and may take longer to
+// finish. If the chain gets rolled back while Process is running, the results
+// are considered invalid and SetStored is not called.
+type ChainSectionProcessorBackend interface {
+	Process(idx uint64) bool
+	GetStored() uint64
+	SetStored(count uint64)
+}
+
+// NewChainSectionProcessor creates a new  ChainSectionProcessor
+func NewChainSectionProcessor(backend ChainSectionProcessorBackend, sectionSize, confirmReq uint64, procWait time.Duration, stop chan struct{}) *ChainSectionProcessor {
+	csp := &ChainSectionProcessor{
+		backend:     backend,
+		sectionSize: sectionSize,
+		confirmReq:  confirmReq,
+		stop:        stop,
+		procWait:    procWait,
+		updateChn:   make(chan uint64, 100),
+		stored:      backend.GetStored(),
+	}
+	go csp.updateLoop()
+	return csp
+}
+
+func (csp *ChainSectionProcessor) updateLoop() {
+	tryUpdate := make(chan struct{}, 1)
+	updating := false
+	var targetCnt uint64
+
+	for {
+		select {
+		case <-csp.stop:
+			return
+		case targetCnt = <-csp.updateChn:
+			if !updating {
+				updating = true
+				tryUpdate <- struct{}{}
```

Comment:
This is very confusing. Why does the code receive a messge on one channel, then send one to another channel that it listens to elsewhere in the same switch statement? Could we not just use a variable?